### ISSUE TITLE
Install sometimes sets incorrect file permissions

### ISSUE THREAD

**Issue by scottkof** [tj/n#592]
# Bug Report

After installation, the `node` executable isn't executable.

## Steps to Reproduce

`$ sudo n 8.9.0`

## Expected Behaviour

```
$ ls -l /usr/local/n/versions/node/8.9.0/bin/
total 34376
-rwxrwxr-x 1 500 500 35197069 Oct 31  2017 node
...
```

## Actual Behaviour

```
$ ls -l /usr/local/n/versions/node/8.9.0/bin/
total 34376
-rwxrwx--- 1 500 500 35197069 Oct 31  2017 node
...
```

## Other Information

This behavior occurs intermittently.

## Configuration Details

```bash
$ n --version
6.0.1

$ command -v node
/usr/local/bin/node

$ node -p process.platform
linux
```

**Comment by shadowspawn** [comment]
I suspect you have a restrictive umask set when the permissions are missing (or something similar which changes the default permissions). e.g.

```sh
$ umask 007
$ n 8.9.0

  installing : node-v8.9.0
       mkdir : /usr/local/n/versions/node/8.9.0
       fetch : https://nodejs.org/dist/v8.9.0/node-v8.9.0-darwin-x64.tar.gz
   installed : v8.9.0 (with npm 5.5.1)

$ ls -l /usr/local/n/versions/node/8.9.0/bin/node
-rwxr-x---  1 jgee  wheel  35828048  1 Nov  2017 /usr/local/n/versions/node/8.9.0/bin/node
```

**Comment by scottkof** [comment]
This issue occurs on hosts with the default umask `0022`. Further, one install on a host can come out just fine, and moments later the next install on the same host can come out with bad perms.

**Comment by shadowspawn** [comment]
Is this happening only when there is a download involved?
Is the download a tarball compressed with gz or with xz?

The two things that happen around the download are the tarball contents are extracted, and some calls are made to `paxctl` if available. The extract is similar to this (I am on a Mac):

```
curl --location https://nodejs.org/dist/v8.9.0/node-v8.9.0-darwin-x64.tar.gz | tar -Jx --strip-components=1
```

There are some differences with whether sudo is involved, from http://manpages.ubuntu.com/manpages/eoan/man1/tar.1.html:

> --no-same-permissions
>              Apply the user's umask when extracting permissions from the  archive  (default  for
>              ordinary users).

> -p, --preserve-permissions, --same-permissions
>              extract information about file permissions (default for superuser)

**Comment by shadowspawn** [comment]
I did some hunting for permission issues with tar extraction, and the closest match I found was this unsolved issue which shares some similarities, but not the intermittent nature:

- https://superuser.com/questions/1482481/tar-does-not-preserve-owner-and-permissions-even-when-invoked-as-root

**Comment by scottkof** [comment]
To answer your questions: it's only happened when downloading gzipped files from nodejs.org.

I've had a chance to look more closely at this, and I've also examined [the n executable](https://github.com/tj/n/blob/master/bin/n). I don't think the issue lies with `n`, I suspect the issue is on the server side of the download.

I didn't mention other weird behavior because I wanted to keep this issue clean. For the record:
- I saw bad file perms on 3 consecutive downloads, and 4 times out of 5. The next day I saw the problem once more but have been unable to reproduce it since then.
- One time during testing I saw a finished install that contained nothing but the `/lib` directory. There should also be `/bin`, `/include`, and `/share`, plus a few text files.

Thanks for taking a look. I'll close this issue.
